pkg/build/buildkit/scaler/downscaler.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 tsuru authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package scaler
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/tsuru/deploy-agent/pkg/build/metadata"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+func StartWorker(clientset *kubernetes.Clientset, podSelector, buildkitStefulset string) {
+	ctx := context.Background()
+
+	go func() {
+		for {
+			err := RunDownscaler(ctx, clientset, podSelector, buildkitStefulset)
+			if err != nil {
+				klog.Errorf("failed to run downscaler tick: %s", err.Error())
+			}
+			time.Sleep(time.Minute * 5)
+		}
+	}()
+}
+
+func RunDownscaler(ctx context.Context, clientset kubernetes.Interface, podSelector, buildkitStefulset string) (err error) {
+	defer func() {
+		recoverErr := recover()
+		if recoverErr != nil {
+			err = fmt.Errorf("panic: %s", recoverErr)
+		}
+	}()
+
+	buildKitPods, err := clientset.CoreV1().Pods("").List(ctx, v1.ListOptions{
+		LabelSelector: podSelector,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	maxEndtimeByNS := map[string]int64{}
+
+	for _, pod := range buildKitPods.Items {
+		if pod.Annotations[metadata.DeployAgentLastBuildEndingTimeLabelKey] == "" {
+			maxEndtimeByNS[pod.Namespace] = -1 // mark that namespace has least one pod of buildkit running
+			continue
+		}
+
+		maxUsage, err := strconv.ParseInt(pod.Annotations[metadata.DeployAgentLastBuildEndingTimeLabelKey], 10, 64)
+		if err != nil {
+			klog.Errorf("failed to parseint: %s", err.Error())
+			continue
+		}
+
+		if maxEndtimeByNS[pod.Namespace] == -1 {
+			continue
+		}
+
+		if maxEndtimeByNS[pod.Namespace] < maxUsage {
+			maxEndtimeByNS[pod.Namespace] = maxUsage
+		}
+	}
+
+	now := time.Now().Unix()
+	gracefulPeriod := int64(60 * 60 * 2) // 2 Hours
```

Comment:
Can we have this as an argument parameter, such as `BuildKitAutoDiscoveryGracefulPeriod`?

Reply:
Yeap, we can!